File: backend/app/schemas/reports.py
from __future__ import annotations

from datetime import date
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class MonthlyUserItem(BaseModel):
    user_id: str = Field(..., description="UUID user en string")
    user_name: str
    month: str = Field(..., pattern=r"^\d{4}-\d{2}$", description="AAAA-MM en UTC")
    hours_planned: float = Field(ge=0)
    hours_confirmed: float = Field(ge=0)
    amount: float = Field(ge=0)


class MonthlyUsersResponse(BaseModel):
    org_id: str
    project_id: Optional[str] = None
    date_from: date
    date_to: date
    items: List[MonthlyUserItem]
    currency: Literal["EUR"] = "EUR"

File: frontend/src/api/reports.ts
export type MonthlyUserItem = {
  user_id: string;
  user_name: string;
  month: string; // YYYY-MM
  hours_planned: number;
  hours_confirmed: number;
  amount: number;
};

export type MonthlyUsersResponse = {
  org_id: string;
  project_id?: string | null;
  date_from: string; // YYYY-MM-DD
  date_to: string; // YYYY-MM-DD
  items: MonthlyUserItem[];
  currency: "EUR";
};

const API_BASE = import.meta.env.VITE_API_BASE ?? "http://localhost:8000";

export async function fetchMonthlyUsers(params: {
  org_id: string;
  project_id?: string;
  date_from: string;
  date_to: string;
}): Promise<MonthlyUsersResponse> {
  const url = new URL("/api/v1/reports/monthly-users", API_BASE);
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
  });
  const r = await fetch(url.toString(), { headers: { Accept: "application/json" } });
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  return r.json();
}

export function exportCSV(params: {
  org_id: string;
  project_id?: string;
  date_from: string;
  date_to: string;
}): void {
  const url = new URL("/api/v1/exports/csv", API_BASE);
  url.searchParams.set("type", "monthly-users");
  Object.entries(params).forEach(([k, v]) => v != null && url.searchParams.set(k, String(v)));
  window.location.href = url.toString();
}
